_posts/2023-04-04-segment-replication.md

```diff
@@ -0,0 +1,241 @@
+
+All operations that affect an OpenSearch index (for example, adding, updating, or removing documents) are routed to one of the index’s primary shards. The primary shard is responsible for validating the operation and subsequently executing it locally. Once the operation has been completed successfully on the primary shard, the operation is forwarded to each of its replica shards in parallel. Each replica shard executes the operation, duplicating the processing performed on the primary shard. When the operation has completed (either successfully or with a failure) on every replica and a response has been received by the primary shard, the primary shard responds to the client. The response includes information about replication success or failure on replica shards. Refer to the following diagram for the document replication process.
+
+TODO: add diagram
+
+## Segment replication
+
+With segment replication, documents are indexed only once on the node that contains the primary shard. The produced segments are then copied to all replicas and are made searchable. The underlying Lucene write-once segmented architecture makes copying segments possible: as documents are updated or deleted, new segments are created but the existing segments are left untouched. Additionally, the cluster only acknowledges index requests once all shards have completed the work. 
+
+Segment replication aims to reduce compute costs and improve indexing throughput at the expense of possible replication delays and increased network usage. Segment replication uses a node-to-node replication method, where segments are copied directly from primary shards to their replicas. This replication method is enabled per index and operates within the OpenSearch cluster model and sharding strategy.
+
+Segment replication is compatible with the storage method chosen for the cluster, supporting both local storage with a node-to-node solution and remote stores. Refer to the following diagram for the segment replication process.
+
+TODO: Add diagram
+
+## Performance
+
+Compared to document replication, segment replication performs better in OpenSearch cluster deployments with low replica counts. For example, segment replication is well-suited for log analytics because, typically, log analytics users maintain only one replica. The low replica count requirement is a limitation of node-to-node (peer-to-peer) replication. For higher replica counts, remote store integration works better. 
+
+With segment replication, you can get the same throughput ingestion performance with 9 nodes in the cluster as you would get with 15 nodes with document replication. During experimental testing, our experimental release users have reported up to 40% higher throughput with segment replication than with document replication for the same cluster setup.
```

Comment:
It's confusing to have sometime node number and sometime %. keep it consistant and convert the node numbers to %

Reply:
Can we confidently say that for any number of nodes, if you reduce the number of nodes by 40%, the throughput will be the same? I think having the number of nodes here is fine. WDYT @nandi-github?

Reply:
It is better to stick to the node numbers. It does not scale at 40% when the nodes increase

Reply:
@anasalkouz As the nodes increase, the performance gain on the indexing lowers due to the copying loads. It will also get limited by the bandwidth of the network interfaces, so we may want to be balanced when we claim the performance gains. We can run a quick test to get numbers. It will be a combination of CPU and network bandwidth on the primary.

Reply:
Another way to say this is that with the same number of nodes you can see up to 40% improvement.